### PR TITLE
BUG: Fix df.loc slice support

### LINKED ISSUE #8856: BUG: loc.__setitem__ over MultiIndex with slice key and step != 1 does strange things
Here's how it looks like:

``` python
In [11]: s = pd.Series(np.arange(10), pd.MultiIndex.from_product(([0, 1], list('abcde'))))

In [12]: s
Out[12]: 
0  a    0
0  b    1
0  c    2
0  d    3
0  e    4
1  a    5
1  b    6
1  c    7
1  d    8
1  e    9
dtype: int64

In [13]: s.iloc[::4] = 100; s
Out[13]: 
0  a    100
0  b    100
0  c    100
0  d    100
0  e    100
1  a      5
1  b      6
1  c      7
1  d      8
1  e      9
dtype: int64

In [14]: s.loc[::4] = 200; s
Out[14]: 
0  a    200
0  b    200
0  c    200
0  d    200
0  e    200
1  a      5
1  b      6
1  c      7
1  d      8
1  e      9
dtype: int64

In [15]: s.ix[::4] = 300; s
Out[15]: 
0  a    300
0  b    300
0  c    300
0  d    300
0  e    300
1  a      5
1  b      6
1  c      7
1  d      8
1  e      9
dtype: int64

```

Regular setitem works fine:

``` python
In [16]: s[::4] = 10000; s
Out[16]: 
0  a    10000
0  b      300
0  c      300
0  d      300
0  e    10000
1  a        5
1  b        6
1  c        7
1  d    10000
1  e        9
dtype: int64

```

I have some ideas about iloc refactoring that will fix that for iloc, but loc/ix will still be affected.

### PR DESCRIPTION
**Pull request by Licht-T**
- [x] closes #8856
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/tests/indexing/test_multiindex.py [hunk 1]
```diff
@@ -239,6 +239,15 @@ def test_iloc_getitem_multiindex(self):
             xp = mi_labels.ix['j'].ix[:, 'j'].ix[0, 0]
         assert rs == xp
 
+        # GH8856
```
Comment (jreback) [inline pandas/tests/indexing/test_multiindex.py]: make separate test

File: pandas/core/indexes/multi.py [hunk 2]
```diff
@@ -2097,7 +2097,9 @@ def _maybe_to_slice(loc):
             mask[loc] = True
             return mask
 
-        if not isinstance(key, tuple):
+        if isinstance(key, slice):
```
Comment (jreback) [inline pandas/core/indexes/multi.py]: can u add some comments here

File: pandas/tests/indexing/test_multiindex.py [hunk 3]
```diff
@@ -278,6 +287,15 @@ def test_loc_multiindex(self):
             xp = mi_int.ix[4]
         tm.assert_frame_equal(rs, xp)
 
+        # GH8856
+        s = pd.Series(np.arange(10),
```
Comment (jreback) [inline pandas/tests/indexing/test_multiindex.py]: is this a duplicate of the other test?

File: pandas/tests/indexing/test_multiindex.py [hunk 4]
```diff
@@ -239,6 +239,15 @@ def test_iloc_getitem_multiindex(self):
             xp = mi_labels.ix['j'].ix[:, 'j'].ix[0, 0]
         assert rs == xp
 
+        # GH8856
+        s = pd.Series(np.arange(10),
+                      pd.MultiIndex.from_product(([0, 1], list('abcde'))))
```
Comment (jreback) [inline pandas/tests/indexing/test_multiindex.py]: can u add several cases using slices

also if other testing of slice u can pull to this section

**Review by jreback — CHANGES_REQUESTED** [review]
whatsnew 0.22

**Comment by Licht-T** [comment]
@jreback Oops, sorry! I overlooked another test.
The `MultiIndex` `loc` slicing work at the first level.
https://github.com/pandas-dev/pandas/blob/master/pandas/tests/test_multilevel.py#L1829
### Input:
```python
s = pd.Series(np.arange(8), pd.MultiIndex.from_product(([0, 1, 2, 3], list('ab'))))
s.loc[::2] = 200
s
```
### Output:
```
0  a    200
   b    200
1  a      2
   b      3
2  a    200
   b    200
3  a      6
   b      7
```

But, this returns the crazy result.
### Input:
```python
index = pd.MultiIndex(levels=[['foo', 'bar', 'baz', 'qux'], ['one', 'two','three']],
                      labels=[[0, 0, 0, 1, 1, 2, 2, 3, 3, 3], [0, 1, 2, 0, 1, 1, 2, 0, 1, 2]],
                      names=['first', 'second'])
frame = pd.DataFrame(np.random.randn(10, 3), index=index, columns=pd.Index(['A', 'B', 'C'], name='exp'))
frame.loc['foo':'bar':4]
```
### Output:
```
exp                  A         B         C
first second
foo   one     0.870785  1.900457 -0.907938
bar   two    -0.290714 -0.211697 -0.957994
```

**Comment by jreback** [comment]
@toobaz if you'd have a look

**Comment by Licht-T** [comment]
I think slicing by string is not good since `slice` seems not supporting string.

File: pandas/tests/indexing/test_multiindex.py (hunk 3 above)
Comment (toobaz) [reply pandas/tests/indexing/test_multiindex.py]: (``.iloc`` vs. ``.loc``)

**Comment by toobaz** [comment]
@Licht-T : I'm confused. Unless I'm missing anything, your changes to code are limited to ``pd.MultiIndex.get_loc``, but in up-to-date master,
- none of the problematic examples in #8856 hits ``pd.MultiIndex.get_loc``
- none of the tests you add addresses any of the problematic examples in #8856 
- none of the tests you add hits ``pd.MultiIndex.get_loc``

... ?

**Comment by jreback** [comment]
closing as stale. ping if you want to update.

**Pull request closed without merging**